File: Parallelism/C/Activity_5.2_prime_parallel.c
/*
* Activity 5.2 Parallel and concurrent programming
*
* Arantza Monique Mercado Moreno
* 2026 - 06 - 03
*
* The sum of all prime numbers smaller than or equal to n (parallel)
*
*/

#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <stdbool.h>    // Library required to use true or false
#include <time.h>

typedef struct {
    int id;
    int start;
    int end;
    // Pointer to the variable where the result will be computed
    unsigned long long * result;
    pthread_mutex_t * mutex_ptr;
} data_t;

// Definition of a pointer to a function
typedef unsigned long long (*primes_function_t)(int, int);

// Function declaration
unsigned long long parallelPrime(int n, int threads);
void * sumRange(void * data);
bool isPrime(int x);
void timeFunction(primes_function_t fun, int n, int threads);
double timespecToSeconds(struct timespec ts_begin, struct timespec ts_end);

int main(int argc, char * argv[])
{
    // Default values
    int n = 10;
    int threads = 1;

    // Read new values from the command line
    if (argc == 3) {
        threads = atoi(argv[2]);
    }
    if (argc > 1) {
        n = atoi(argv[1]);
    }

    printf("=== PARALLEL VERSION ===\n");
    timeFunction(parallelPrime, n, threads);

    return 0;
}

// Function to compute the sum of all the prime numbers less than o equal to n.
// This version splits the calculation into multiple parts.
unsigned long long parallelPrime(int n, int threads)
{
    int rangeSize = n / threads;
    int remainder = n % threads;

    // Create the array of thread id's
    pthread_t tids[threads];
    
    // Create an array of structures
    data_t thread_data[threads];

    // Create the common mutex lock
    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;

    int status = 0;

    // Variable to store the sum of primes computed in parallel
    unsigned long long resultParallel = 0;

    // Loop to call the functions as many times as threads
    for (int i=0; i < threads; i++) {
        thread_data[i].id = i;
        thread_data[i].start = i * rangeSize + 1;
        thread_data[i].end = (i + 1) * rangeSize;
        thread_data[i].result = &resultParallel;
        thread_data[i].mutex_ptr = &mutex;

        if (i == threads - 1){
            thread_data[i].end +=  remainder;
        }

        printf("ID: %d [%d, %d]\n", thread_data[i].id, thread_data[i].start, thread_data[i].end);

        status = pthread_create(&tids[i], NULL, sumRange, &thread_data[i]);
        if (status == -1)
        {
            perror("ERROR: pthread_create");
        }
    }

    for (int i = 0; i<threads; i++) {
        status = pthread_join(tids[i], NULL);
        if (status == -1)
        {
            perror("ERROR: pthread_join");
        }
    }

    return resultParallel;
}

// Compute the sum of prime numbers in a range
void * sumRange(void * data)
{
    data_t * info = data;
    
    // Cast the pointer into the desired type
    data_t * localData = (data_t *) data;
    unsigned long long result = 0;

    for (int i = localData -> start; i <= localData->end; i++) {
        if (isPrime(i) == true){        // Calls another function to identify if it is a prime number
            result += i;
        }
    }

    // Lock access to the shared memory, before modifying it
    pthread_mutex_lock(info->mutex_ptr);

    // Update the shared variable
    (*localData->result) += result;
    pthread_mutex_unlock(localData -> mutex_ptr);

    // Return the address in heap
    pthread_exit(NULL);

    return 0;
}

// Function that identifies whether a number is prime or not
bool isPrime(int x)
{
     if (x < 2){
        return false;
    }
    if (x == 2) {
        return true;
    }
    for (int i = 2; i * i <= x; i++){
        if (x % i == 0){
            return false;
        }
    }
    return true;

}

/*
 * Wrapper function that measures the time taken by another function to execute
 * Receives a pointer to the function, and two arguments
 */
void timeFunction(primes_function_t fun, int n, int threads)
{
    // Variables to measure time (this one measures cpu time)
    // https://www.geeksforgeeks.org/cpp/how-to-measure-time-taken-by-a-program-in-c/
    /*
    clock_t start;
    clock_t end;
    double cpu_time_used;

    start = clock();
    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;
    */


    // https://levelup.gitconnected.com/8-ways-to-measure-execution-time-in-c-c-48634458d0f9
    struct timespec ts_wall_begin, ts_wall_end;
    struct timespec ts_cpu_begin, ts_cpu_end;
    double elapsed_wall;
    double elapsed_cpu;

    clock_gettime(CLOCK_REALTIME, &ts_wall_begin);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &ts_cpu_begin);

    // Call the function
    unsigned long long result = fun(n, threads);

    clock_gettime(CLOCK_REALTIME, &ts_wall_end);
    clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &ts_cpu_end);

    // Compute the times elapsed
    elapsed_wall = timespecToSeconds(ts_wall_begin, ts_wall_end);
    elapsed_cpu = timespecToSeconds(ts_cpu_begin, ts_cpu_end);

    printf("N: %d | Result: %llu | Time: %lf | CPU: %lf\n", n, result, elapsed_wall, elapsed_cpu);
}

// Compute the total amount of seconds from a timespec
double timespecToSeconds(struct timespec ts_begin, struct timespec ts_end)
{
    long seconds;
    long nanoseconds;

    seconds = ts_end.tv_sec - ts_begin.tv_sec;
    nanoseconds = ts_end.tv_nsec - ts_begin.tv_nsec;
    return seconds + nanoseconds * 1e-9;
}

File: Parallelism/Elixir/Activity_5.2_prime_parallel.exs
# Activity 5.2 Parallel and concurrent programming

# Arantza Monique Mercado Moreno
# 2026 - 06 - 03

# Sum of all the prime numbers equal to or less than n (parallel)

defmodule Primes do

  # Function that identifies whether the number is prime or not
  # Base cases
  def is_prime?(x) when x < 2, do: false
  def is_prime?(2), do: true
  # Recursive step: check divisors from 2 to √x
  def is_prime?(x), do: do_is_prime?(x, 2)

  # Base cases
  defp do_is_prime?(x, i) when i * i > x, do: true
  defp do_is_prime?(x, i) when rem(x, i) == 0, do: false
  # Recursive step
  defp do_is_prime?(x, i), do: do_is_prime?(x, i + 1)

  # Sum all the prime numbers within a range
  # Receives a tuple
  def sum_range({start, stop}), do: do_sum_range(start, stop, 0)
  # Base case
  defp do_sum_range(i, stop, res) when i > stop, do: res
  # Recursive step
  defp do_sum_range(i, stop, res) do
      if is_prime?(i) do
          do_sum_range(i + 1, stop, res + i)
      else
          do_sum_range(i + 1, stop, res)
      end
  end

  # Compute the sum using multiple threads
  # The second parameter has as default value the number of cores
  def parallel_primes(n, tasks \\ System.schedulers) do
    step = ceil(n / tasks)
    starts = [1 | Enum.to_list(step+1..n//step)]
    finishes = Enum.map(starts, fn start -> min(start + step - 1, n) end)

    # Using a custom made pmap function
    Enum.zip(starts, finishes)
    |> IO.inspect()
    |> Enum.map(&Task.async(fn -> sum_range(&1) end))
    |> Enum.map(&Task.await(&1, :infinity))
    |> Enum.reduce(0, &(&1 + &2))
  end

  # Function to get the time taken by the function to run
  # Call it this way:
  # Primes.measure_time(&Primes.parallel_primes/2, [10, 4])
  def measure_time(function, parameters) do
    {time, _result} = :timer.tc(function, parameters)
    # Divide by a million, to get the time in seconds
    time / 1_000_000
  end

end
